aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+}
+
+static Tensor& linalg_eig_make_complex_eigenvectors(Tensor& complex_vectors, const Tensor& complex_values, const Tensor& real_vectors) {
+  // These asserts make explicit the requirements on tensors for 'linalg_eig_make_complex_eigenvectors_impl'
+  TORCH_INTERNAL_ASSERT(complex_vectors.device() == at::kCPU);
+  TORCH_INTERNAL_ASSERT(complex_values.device() == at::kCPU);
+  TORCH_INTERNAL_ASSERT(real_vectors.device() == at::kCPU);
+
+  TORCH_INTERNAL_ASSERT(complex_vectors.is_complex());
+  TORCH_INTERNAL_ASSERT(complex_values.is_complex());
+  TORCH_INTERNAL_ASSERT(real_vectors.is_floating_point());
+
+  TORCH_INTERNAL_ASSERT(complex_vectors.transpose(-2, -1).is_contiguous());
+  TORCH_INTERNAL_ASSERT(complex_values.is_contiguous());
+  TORCH_INTERNAL_ASSERT(real_vectors.transpose(-2, -1).is_contiguous());
+
+  AT_DISPATCH_FLOATING_TYPES(real_vectors.scalar_type(), "linalg_eig_make_complex_vector", [&]{
+    linalg_eig_make_complex_eigenvectors_impl<scalar_t>(complex_vectors, complex_values, real_vectors);
+  });
+  return complex_vectors;
+}
+
+DEFINE_DISPATCH(linalg_eig_stub);
+
+std::tuple<Tensor&, Tensor&> linalg_eig_out_info(const Tensor& input, Tensor& values, Tensor& vectors, Tensor& infos, bool compute_eigenvectors) {
+  // MAGMA doesn't have GPU interface for GEEV routine, it requires inputs to be on CPU
+  // therefore we create all intermediate tensors on CPU
+  auto options = input.options().device(at::kCPU);
+
+  // These internal asserts make explicit the assumptions in the implementation
+  // Error check with the actual error messages are done on the higher level of the hierarchy of calls
+  TORCH_INTERNAL_ASSERT(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT(input.size(-2) == input.size(-1));
+
+  // for real-valued 'input', eigenvalues can be real-valued or complex-valued
+  TORCH_INTERNAL_ASSERT((toComplexType(input.scalar_type()) == values.scalar_type()) || (input.scalar_type() == values.scalar_type()));
+  TORCH_INTERNAL_ASSERT(values.device() == at::kCPU);
+
+  // for real-valued 'input', eigenvectors can be real-valued or complex-valued
+  if (compute_eigenvectors) {
+    TORCH_INTERNAL_ASSERT((toComplexType(input.scalar_type()) == vectors.scalar_type()) || (input.scalar_type() == vectors.scalar_type()));
+    TORCH_INTERNAL_ASSERT(vectors.device() == at::kCPU);
+  }
+
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.device() == at::kCPU);
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
+  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+
+  // if 'vectors' has no elements we can modify it
+  if (vectors.numel() == 0 && compute_eigenvectors) {
+    vectors.resize_(input.sizes(), MemoryFormat::Contiguous);
+    vectors.transpose_(-2, -1);  // make 'vectors' to have Fortran contiguous memory layout
+  }
+
+  // if 'values' has no elements we can modify it
+  auto values_shape = IntArrayRef(input.sizes().data(), input.dim()-1);  // input.shape[:-1]
```

Comment:
How does this interact with the internal asserts later?

If values can come from users, then shouldn't the later internal asserts be torch checks? If values only comes internally, shouldn't its properties just be asserted?

Reply:
This function has two modes of operation:

- if passed "out" tensors are empty then they can be resized, this makes calling this function from `linalg_eig` and `linalg_eig_out` simpler.
- if non-empty tensors are passed they must be in lapack compatible format: column-major memory layout, contiguous, and of correct shape.

Possible resizing of passed by user tensors happen outside of this function in `linalg_eig_out`. This avoids recursion and separates the scope of functions.